tensorflow_probability/python/experimental/mcmc/with_reductions_test.py

```diff
@@ -0,0 +1,491 @@
+
+
+class TestTransitionKernel(tfp.mcmc.TransitionKernel):
+  """Fake deterministic Transition Kernel"""
+
+  def __init__(self, shape=(), target_log_prob_fn=None, is_calibrated=True):
+    self._is_calibrated = is_calibrated
+    self._shape = shape
+    # for composition purposes
+    self.parameters = dict(
+        target_log_prob_fn=target_log_prob_fn)
+
+  def one_step(self, current_state, previous_kernel_results, seed=None):
+    return (current_state + tf.ones(self._shape),
+            TestTransitionKernelResults(
+                counter_1=previous_kernel_results.counter_1 + 1,
+                counter_2=previous_kernel_results.counter_2 + 2))
+
+  def bootstrap_results(self, current_state):
+    return TestTransitionKernelResults(
+        counter_1=tf.zeros(()),
+        counter_2=tf.zeros(()))
+
+  @property
+  def is_calibrated(self):
+    return self._is_calibrated
+
+
+@test_util.test_all_tf_execution_regimes
+class WithReductionsTest(test_util.TestCase):
+
+  def test_simple_operation(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    new_sample, kernel_results = reducer_kernel.one_step(0.,)
+    new_sample, kernel_results = self.evaluate([
+        new_sample, kernel_results])
+    self.assertEqual(kernel_results.streaming_calculations, 1)
+    self.assertEqual(new_sample, 1)
+    self.assertEqual(kernel_results.inner_results.counter_1, 1)
+    self.assertEqual(kernel_results.inner_results.counter_2, 2)
+
+  def test_boostrap_results(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    pkr = reducer_kernel.bootstrap_results(9.)
+    inner_pkr = self.evaluate(pkr.inner_results)
+    self.assertEqual(pkr.streaming_calculations, None)
+    self.assertEqual(inner_pkr.counter_1, 0)
+    self.assertEqual(inner_pkr.counter_2, 0)
+
+  def test_is_calibrated(self):
+    fake_calibrated_kernel = TestTransitionKernel()
+    fake_uncalibrated_kernel = TestTransitionKernel(is_calibrated=False)
+    fake_reducer = TestReducer()
+    calibrated_reducer_kernel = WithReductions(
+        inner_kernel=fake_calibrated_kernel,
+        reducers=fake_reducer,
+    )
+    uncalibrated_reducer_kernel = WithReductions(
+        inner_kernel=fake_uncalibrated_kernel,
+        reducers=fake_reducer,
+    )
+    self.assertTrue(calibrated_reducer_kernel.is_calibrated)
+    self.assertFalse(uncalibrated_reducer_kernel.is_calibrated)
+
+  def test_tf_while(self):
+    fake_kernel = TestTransitionKernel()
+    fake_reducer = TestReducer()
+    reducer_kernel = WithReductions(
+        inner_kernel=fake_kernel,
+        reducers=fake_reducer,
+    )
+    initial_sample, initial_kernel_results = reducer_kernel.one_step(0.,)
```

Comment:
Test just bootstrapping here, without taking a step.  You might experience an interesting problem, whose solution will be to change the `initialize` of the `TestReducer` to produce a Tensor of the same shape as `initial_sample`, instead of `None`.

The reason to test like this is that the drivers generally will not take a step outside of the `tf.while_loop`.